### DISCUSSION TITLE
How can I add gRPC decorator?

### DISCUSSION THREAD

**Discussion by hi-hj** [line/armeria#5902] [Q&A]
How can I add **custom Decorator to gRPC** API?
I tried to replace DecoratingHttpServiceFunction interface to DecorationgRpcServiceFunction, But I can not find appropriate example(https://armeria.dev/docs/server-decorator/)

I already added Decorator to REST API (annotated service) using Decorator annotation. and it works.
```
@Decorator(HandlerHttpFunction::class)
@Component
class RestController(
```

```
class HandlerHttpFunction : DecoratingHttpServiceFunction {
    override fun serve(delegate: HttpService, ctx: ServiceRequestContext, req: HttpRequest): HttpResponse {
        val testId= ctx.pathParams()["testID"] ?: throw IllegalArgumentException(ExceptionMessage.EMPTY_TEST_ID)
        ServiceRequestContext.current().setAttr(TEST_ID_KEY, testId)

        return delegate.serve(ctx, req)
    }
}
```

```
sb.annotatedService(RestController, AuthService.newDecorator(integratedJwtAuthorizer))
```
And here is my gRPC code. How can I add same or similar function to gRPC like REST?
Could you share some example or guide?

```
@Component
class GrpcController(
    private val grpcUseCase: GrpcUseCase
) : GrpcServiceGrpcKt.GrpcServiceCoroutineImplBase() {

```
        sb.service(
            GrpcService.builder()
                .addService(GrpcController)
                .build(),
            LoggingService.newDecorator()
        )

**Comment by ikhoon** [comment]
> I tried to replace DecoratingHttpServiceFunction interface to DecorationgRpcServiceFunction

`DecorationgRpcServiceFunction` exists for Thrift because it does not have a cross-cutting concept in Thrift.

If we want to use an HTTP level decorator, `DecoratingHttpServiceFunction` will work for your gRPC services.

>     sb.service(
>        GrpcService.builder()
>            .addService(GrpcController)
>            .build(),
>        LoggingService.newDecorator()
>    )

Doesn't this example work? `LoggingService` should work gRPC requests.

**Reply by hi-hj** [reply]
Thank you for reply, I understand that `DecoratingHttpServiceFunction` can cover REST and gRPC API, and it works 😄 
I think it was problem of my custom decorator handler (`HandlerHttpFunction`)

My code can handle REST API that `testID` is in path variable.
```
val testId= ctx.pathParams()["testID"] ?: throw IllegalArgumentException(ExceptionMessage.EMPTY_TEST_ID)
```
But, I have no idea how to handle gRPC API that contains `testId` in message..
If I find solution, I'll share it on this discussion 👍
